src/core/umount.c

```diff
@@ -382,6 +384,82 @@ static bool nonunmountable_path(const char *path) {
                 || path_startswith(path, "/run/initramfs");
 }
 
+static int remount_with_timeout(MountPoint *m, char *options, int *n_failed, bool log_error) {
+        pid_t pid;
+        int r;
+
+        BLOCK_SIGNALS(SIGCHLD);
+
+        /* Due to the possiblity of a remount operation hanging, we
+         * fork a child process and set a timeout. If the timeout
+         * lapses, the assumption is that that particular remount
+         * failed. */
+        pid = fork();
+        if (pid < 0)
+                return log_error_errno(errno, "Failed to fork: %m");
+
+        if (pid == 0) {
+                log_info("Remounting '%s' read-only in with options '%s'.", m->path, options);
+
+                /* Start the mount operation here in the child */
+                r = mount(NULL, m->path, NULL, MS_REMOUNT|MS_RDONLY, options);
+                if (r < 0) {
+                        if (log_error)
+                                log_error_errno(errno, "Failed to remount '%s' read-only: %m", m->path);
+                }
+                _exit(r < 0 ? EXIT_FAILURE : EXIT_SUCCESS);
+        }
+
+        r = wait_for_terminate_with_timeout(pid, DEFAULT_TIMEOUT_USEC);
+        if (r < 0 && r == -ETIMEDOUT) {
```

Comment:
if r is -ETIMEDOUT then it's always negative, no need to check `r < 0` too...

that sad, i think it would be good to also log in the other error cases:

```
r = wait_for_terminate_with_timeout(…)
if (r == -ETIMEDOUT) {
        …
}
if (r < 0)
        log_error_errno(r, …);
```